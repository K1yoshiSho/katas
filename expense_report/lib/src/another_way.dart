import 'package:approval_tests/approval_tests.dart';

class Expense {
  final ExpenseType type;
  final int amount;

  const Expense({
    required this.type,
    this.amount = 0,
  });
}

enum ExpenseType {
  dinner,
  breakfast,
  carRental,
  lunch,
}

extension ExpenseExtension on Expense {
  Service get service => switch (type) {
        ExpenseType.breakfast => Breakfast(amount: amount),
        ExpenseType.lunch => Lunch(amount: amount),
        ExpenseType.dinner => Dinner(amount: amount),
        ExpenseType.carRental => CarRental(amount: amount),
      };
}

abstract class Service {
  final int amount;

  const Service({required this.amount});

  String get name;
}

mixin MealExpense on Service {
  int get expenseLimit;

  bool get isOverLimit => amount > expenseLimit;
}

class Breakfast extends Service with MealExpense {
  const Breakfast({required super.amount});

  @override
  int get expenseLimit => 1000;

  @override
  String get name => 'Breakfast';
}

class CarRental extends Service {
  const CarRental({required super.amount});

  @override
  String get name => 'Car Rental';
}

class Dinner extends Service with MealExpense {
  const Dinner({required super.amount});

  @override
  int get expenseLimit => 5000;

  @override
  String get name => 'Dinner';
}

class Lunch extends Service with MealExpense {
  const Lunch({required super.amount});

  @override
  int get expenseLimit => 2000;

  @override
  String get name => 'Lunch';
}

class ExpenseReport {
  void printReport(List<Expense> expenses, {required DateTime date}) {
    _writeHeader(date);
    _writeExpenses(expenses);
    _writeTotals(expenses);
  }

  void _writeHeader(DateTime date) {
    log('Expense Report: $date');
  }

  void _writeExpenses(List<Expense> expenses) {
    for (final expense in expenses) {
      _writeExpenseLine(expense);
    }
  }

  void _writeExpenseLine(Expense expense) {
    final service = expense.service;
    final isOverLimit = service is MealExpense && service.isOverLimit;
    log('${service.name}\t${expense.amount}\t${isOverLimit ? 'X' : ' '}');
  }

  void _writeTotals(List<Expense> expenses) {
    final mealExpenses = _calculateMealExpenses(expenses);
    final totalExpenses =
        expenses.fold(0, (total, expense) => total + expense.amount);
    log('Meal Expenses: $mealExpenses');
    log('Total Expenses: $totalExpenses');
  }

  int _calculateMealExpenses(List<Expense> expenses) => expenses
      .where((expense) => expense.service is MealExpense)
      .fold(0, (total, expense) => total + expense.amount);

  void log(Object? object) => ApprovalLogger.log(object.toString());
}
