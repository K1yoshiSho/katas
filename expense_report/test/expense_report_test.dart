import 'package:approval_tests/approval_tests.dart';
import 'package:expense_report/src/another_way.dart';
import 'package:test/test.dart';

const input = [
  Expense(type: ExpenseType.dinner, amount: 1),
  Expense(type: ExpenseType.breakfast, amount: 2),
  Expense(type: ExpenseType.carRental, amount: 4),
  Expense(type: ExpenseType.dinner, amount: 5000),
  Expense(type: ExpenseType.dinner, amount: 5001),
  Expense(type: ExpenseType.breakfast, amount: 1000),
  Expense(type: ExpenseType.breakfast, amount: 1001),
];
void main() {
  late TestableExpenseReport expenseReport;
  final DateTime date = DateTime(2022, 9);

  group('ExpenseReport', () {
    setUp(() {
      expenseReport = TestableExpenseReport();
    });

    test('printReport', () {
      expenseReport.printReport(input, date: date);
      Approvals.verify(
        expenseReport.logs.join('\n'),
        options: const Options(
          comparator: IDEComparator(ide: ComparatorIDE.visualStudioCode),
        ),
      );
    });
  });
}

final class TestableExpenseReport extends ExpenseReport {
  final List<String> logs = [];
  @override
  void log(Object? object) {
    logs.add(object.toString());
  }
}
