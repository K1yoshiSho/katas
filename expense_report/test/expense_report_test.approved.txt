Expense Report: 2022-09-01 00:00:00.000
Dinner	1	 
Breakfast	2	 
Car Rental	4	 
Dinner	5000	 
Dinner	5001	X
Breakfast	1000	 
Breakfast	1001	X
Lunch	1001	 
Lunch	2001	X
Meal Expenses: 15007
Total Expenses: 15011